oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/WriteAfterRecoveryTest.java

```diff
@@ -0,0 +1,661 @@
+        // add a node
+        try {
+            merge(ns2, createChild("/a/b/c/d/e/f/g"), false);
+            fail("merge must fail");
+        } catch (CommitFailedException e) {
+            assertThat(e.getMessage(), containsString("already deleted"));
+        }
+        // can set a property on a node where recovery reverted the delete ...
+        merge(ns2, setProperty("/a/b", "p", "v"), false);
+        // ... but cannot set a property on a node deleted after recovery
+        try {
+            merge(ns2, setProperty("/a/b/c/d/e/f", "p", "v"), false);
+            fail("merge must fail");
+        } catch (CommitFailedException e) {
+            assertThat(e.getMessage(), containsString("already deleted"));
+        }
+
+        Set<Revision> inconsistent = checkConsistency(ns2, "/a/b", "/a/b/c", "/a/b/c/d", "/a/b/c/d/e", "/a/b/c/d/e/f", "/a/b/c/d/e/f/g");
+        assertEquals(lateWriteRevisions, inconsistent);
+    }
+
+    @Test
+    public void propertyChanged() throws Exception {
+        merge(ns1, compose(
+                createChild("/a/b"),
+                setProperty("/a/b", "p", "u")
+        ), true);
+
+        List<UpdateOp> failed = new ArrayList<>();
+        store.addListener(filter(failed, setPropertyOps("p")));
+        store.fail().after(0).eternally();
+        try {
+            merge(ns1, compose(
+                    setProperty("/a/b", "p", "v"),
+                    createChild("/a/b/c")
+            ), false);
+            fail("merge must fail");
+        } catch (CommitFailedException e) {
+            // expected
+        }
+        disposeQuietly(ns1);
+        store.fail().never();
+
+        // wait until lease expires
+        clockWait(DEFAULT_LEASE_DURATION_MILLIS);
+
+        DocumentNodeStore ns2 = createNodeStore(2);
+        assertTrue(ns2.getLastRevRecoveryAgent().isRecoveryNeeded());
+        assertEquals(0, ns2.getLastRevRecoveryAgent().recover(1));
+
+        // 'late write'
+        store.createOrUpdate(NODES, failed);
+        Set<Revision> lateWriteRevisions = getRevisions(failed);
+
+        // revive clusterId 1
+        ns1 = createNodeStore(1);
+        merge(ns1, createChild("/d"), true);
+        assertEquals(stringProperty("p", "u"), getNode(ns1.getRoot(), "/a/b").getProperty("p"));
+        NodeDocument b = store.find(NODES, getIdFromPath("/a/b"));
+        assertNotNull(b);
+        assertNotNull(b.getNodeAtRevision(ns1, ns1.getHeadRevision(), null).getProperty("p"));
+
+        ns2.runBackgroundOperations();
+        assertTrue(getNode(ns2.getRoot(), "/d").exists());
+        assertEquals(stringProperty("p", "u"), getNode(ns2.getRoot(), "/a/b").getProperty("p"));
+
+        // current implementation does not fail with conflict when property
+        // is overwritten, but consistency check can still detect it
+        Set<Revision> inconsistent = checkConsistency(ns2, "/a/b", "/a/b/c");
+        assertEquals(lateWriteRevisions, inconsistent);
+    }
```

Comment:
```suggestion
        // by making another change on a property with an inconsistency,
        // that inconsistency then becomes unnoticeable going forward
        assertEquals(1, checkConsistency(ns2, "/a/b").size());
        merge(ns2, setProperty("/a/b", "p", "v2"), true);
        assertEquals(0, checkConsistency(ns2, "/a/b").size());
    }
```

It might be obvious, but perhaps we can add it in one place, that if a property with an inconsistency is further updated, that inconsistency "disappears".

Reply:
PS: there might be exotic cases that one could construct where a property changes as a result of such an overwrite when reading the node via a checkpoint (it is fine for head though)

Reply:
And now I remember another test idea : when making a change on a *different* property on the same node, the original property state also changes - that might be a useful test to add perhaps. 
This could also be further extended with a test where a node was removed via a late write, at which point it would purely be an inconsistency. However, a later modification on a property on that removed node would actually immediately remove the node.

Reply:
Added all those suggestions, except for the exotic case, which I didn't understand. Can you please explain in more detail what you have in mind?

Reply:
With exotic I was refering to a case where someone would have a reference on an inconsistent property value - that normally gets filtered out, but only until _lastRev on the document changes because of an unrelated change (which is similar to the test of setting an unrelated property that makes the inconsistency disappear). If someone has a checkpoint to the document between the inconsistency and a new update, then the new update changes the property value for the one reading via the checkpoint. Eg as below:

```
        String checkpoint = ns1.checkpoint(Long.MAX_VALUE);
        assertEquals(stringProperty("p", "u"), getNode(ns1.retrieve(checkpoint), "/a/b").getProperty("p"));
        merge(ns1, createChild("/a/b/e"), true);
        // invalidation is required here as otherwise it reads an old node state that doesn't trigger a failure in below assertEquals
        ns1.getNodeCache().invalidateAll();
        assertEquals(stringProperty("p", "u"), getNode(ns1.retrieve(checkpoint), "/a/b").getProperty("p"));
```

Reply:
OK, thanks. I added similar asserts to test propertyChanged().